### DISCUSSION TITLE
Orchard Core Module vs Razor Class Library

### DISCUSSION THREAD

**Discussion by DrewBrasher** [OrchardCMS/OrchardCore#11355] [Q&A] [answered]
I am planning to rebuild some asp.net web forms apps in .net core mvc.  I would like to build each feature as separate components that can be added to the new app. The Orchard Core framework seems perfect for this but it also seems like it could be done with razor class libraries also.  What would some advantages be to using Orchard Core Modules instead of Razor Class Libraries?

**Comment by hishamco** [comment]
OC modules are using RCL, please check the project SDK its `Microsoft.NET.Sdk.Razor`, so you can start creating your MVC views or RazorPages nothing prevent you to create your awesome module in the methodology you like

**Reply by DrewBrasher** [reply]
What I'm wondering is what are the benefits to making a RCL an Orchard Module instead of just a plain RCL?

**Reply by Skrypt** [reply]
Yes, it could be a "Core" library that you can use in all your modules by reference. Though, I think you would need to use the /shared folder for your views.

**Reply by ns8482e** [reply]
you can't make RCL an orchard module if it has `Areas` folder https://github.com/OrchardCMS/OrchardCore/issues/4247

**Reply by Skrypt** [reply]
Areas never worked with OrchardCore modules. That's an issue that could be fixed and that I reported a while ago.
So, the limitation is about our implementation details but this should normally work.

The reason why it doesn't work is that a module is an Area and we didn't though about sub-Areas.

**Reply by ns8482e** [reply]
correct  - when adding reference of RCL having areas to  OC module and OC module itself being an `area` RCL areas becomes sub-areas and OC doesn't support sub areas yet

**Comment by ns8482e** [comment, marked as answer]
Using Orchard Core Module
- With Saas recipe, you can get admin experience out of the box.
- Supports multi-tenancy out of the box
- Module represent an area - you don't need to create `Areas` folder in project for separate features.
- In single module you can define more than one `Features` with `[Feature]` attribute that you can enable/disable separately.
- All `cshtml` source files are precompiled but are also embedded into assembly.
- All assets/contents are embedded into assembly
- You can easily add support of shapes/liquid/placement if use DisplayManagement.

Using RCL
- You have to build feature management.
- No support for multi-tenancy you need build yourself or look for another library
- You need to place all controllers explicitly in project under `Areas/feature` folder
- In single project you can define more than one `Features` but requires to define explicitly under `/Areas/FeatureX`, `Areas/FeatureY` 
- All cshtml are pre-complied .
- All assets/contents are bundled as content in nuget package.

**Reply by DrewBrasher** [reply]
Thank you! This is exactly what I was looking for. I had started building two simple test apps using each method to see what the differences were. The two benefits I had seen so far were that OC Module looked cleaner and was automatically mapped. My test apps are here just in case anyone is curious: https://github.com/DrewBrasher/RCLvsOCModule
